test/es-module/test-esm-loader-hooks.mjs

```diff
@@ -0,0 +1,73 @@
+// Flags: --expose-internals
+import { mustCall } from '../common/index.mjs';
+import esmLoaderModule from 'internal/modules/esm/loader';
+import assert from 'assert';
+
+const { ESMLoader } = esmLoaderModule;
+
+{
+  const esmLoader = new ESMLoader();
+
+  const originalSpecifier = 'foo/bar';
+  const importAssertions = { type: 'json' };
+  const parentURL = 'file:///entrypoint.js';
+  const resolvedURL = 'file:///foo/bar.js';
+  const suggestedFormat = 'test';
+
+  function resolve(specifier, context, defaultResolve) {
+    assert.strictEqual(specifier, originalSpecifier);
+    assert.deepStrictEqual(Object.keys(context), [
+      'conditions',
+      'importAssertions',
+      'parentURL',
+    ]);
+    assert.ok(Array.isArray(context.conditions));
+    assert.strictEqual(context.importAssertions, importAssertions);
+    assert.strictEqual(context.parentURL, parentURL);
+    assert.strictEqual(typeof defaultResolve, 'function');
+
+    return {
+      format: suggestedFormat,
+      url: resolvedURL,
+    };
+  }
+
+  function load(resolvedURL, context, defaultLoad) {
+    assert.strictEqual(resolvedURL, resolvedURL);
+    assert.ok(new URL(resolvedURL));
+    assert.deepStrictEqual(Object.keys(context), [
+      'format',
+      'importAssertions',
+    ]);
```

Comment:
Can you make it clear somewhere that what you’re testing in this file is the signatures of the hooks? Like maybe name the file something like `test-esm-loader-hooks-signatures.mjs`, and add some comments near these assertions that what we’re doing here is testing that the expected arguments/properties exist and that unexpected ones _don’t_ exist?

And I assume that’s all you’re aiming to test in this file?

Reply:
> Can you make it clear somewhere that what you’re testing in this file is the signatures of the hooks? […] add some comments near these assertions that what we’re doing here is testing that the expected arguments/properties exist and that unexpected ones don’t exist?

Sure!

> And I assume that’s all you’re aiming to test in this file?

No, I'm planning (in the near future) to add more tests for hooks themselves in this file (hence the generic file name). Those aren't relevant to this PR, so I didn't add them yet.

Reply:
It’s much easier to debug failing tests when each test file is as close as possible to one test per file, since you can re-run a test file like `./node test/es-module/test-esm-loader-hooks.mjs` etc. The loader hooks API is large enough that I can’t imagine testing all of it in one file; you could make a subfolder for it perhaps.